llama_index/vector_stores/google/generativeai/genai_extension.py

```diff
@@ -0,0 +1,534 @@
+                parent=doc_name,
+                chunk=genai.Chunk(
+                    data=genai.ChunkData(string_value=text),
+                    custom_metadata=_convertToMetadata(metadata),
+                ),
+            )
+        )
+    return created_chunks
+
+
+"""Restore this after b/307125429 is fixed.
+operation = client.batch_create_chunks(
+    genai.BatchCreateChunksRequest(
+        parent=doc_name,
+        requests=[
+            genai.CreateChunkRequest(
+                parent=doc_name,
+                chunk=genai.Chunk(
+                    data=genai.ChunkData(string_value=text),
+                    custom_metadata=_convertToMetadata(metadata),
+                ),
+            )
+            for text, metadata in zip(texts, metadatas)
+        ]
+    )
+)
+
+while not operation.done():
+    _logger.info("Waiting on batch_create_chunk operation")
+response = cast(genai.BatchCreateChunksResponse, operation.result())
+return [chunk for chunk in response.chunks]
+"""
+
+
+def delete_chunk(
+    *,
+    corpus_id: str,
+    document_id: str,
+    chunk_id: str,
+    client: Optional[genai.RetrieverServiceClient] = None,
+) -> None:
+    if client is None:
+        client = _default_retriever
+    client.delete_chunk(
+        genai.DeleteChunkRequest(
+            name=str(
+                EntityName(
+                    corpus_id=corpus_id, document_id=document_id, chunk_id=chunk_id
+                )
+            )
+        )
+    )
+
+
+def query_corpus(
+    *,
+    corpus_id: str,
+    query: str,
+    k: int = 4,
+    filter: Optional[Dict[str, Any]] = None,
+    client: Optional[genai.RetrieverServiceClient] = None,
+) -> List[genai.RelevantChunk]:
+    if client is None:
+        client = _default_retriever
+    response = client.query_corpus(
+        genai.QueryCorpusRequest(
+            name=str(EntityName(corpus_id=corpus_id)),
+            query=query,
+            metadata_filters=_convertFilter(filter),
+            results_count=k,
+        )
+    )
+    return list(response.relevant_chunks)
+
+
+def query_document(
+    *,
+    corpus_id: str,
+    document_id: str,
+    query: str,
+    k: int = 4,
+    filter: Optional[Dict[str, Any]] = None,
+    client: Optional[genai.RetrieverServiceClient] = None,
+) -> List[genai.RelevantChunk]:
+    if client is None:
+        client = _default_retriever
+    response = client.query_document(
+        genai.QueryDocumentRequest(
+            name=str(EntityName(corpus_id=corpus_id, document_id=document_id)),
+            query=query,
+            metadata_filters=_convertFilter(filter),
+            results_count=k,
+        )
+    )
+    return list(response.relevant_chunks)
+
+
+@dataclass
+class Passage:
+    text: str
+    ids: List[str]
+
+
+@dataclass
+class TextAnswer:
+    answer: str
+    attributed_passages: List[Passage]
+    answerable_probability: float
+
+
+def generate_text_answer(
+    *,
+    prompt: str,
+    passages: List[str],
```

Comment:
Do we also want to support specifying a corpus ID and/or document ID?

Reply:
Yep, that's fine to do it later. The new endpoint supports filtering and results_count now (see CorpusSource and DocumentSource).